<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Manual &mdash; OpenSMOG3SPN2  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Welcome to OpenSMOG3SPN2’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            OpenSMOG3SPN2
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Manual</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-OpenSMOG3SPN2.forcefields.cg_model">OpenSMOG3SPN2.forcefields</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel"><code class="docutils literal notranslate"><span class="pre">CGModel</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.add_plumed"><code class="docutils literal notranslate"><span class="pre">CGModel.add_plumed()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.add_reporters"><code class="docutils literal notranslate"><span class="pre">CGModel.add_reporters()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.append_mol"><code class="docutils literal notranslate"><span class="pre">CGModel.append_mol()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.atoms_to_pdb"><code class="docutils literal notranslate"><span class="pre">CGModel.atoms_to_pdb()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.create_system"><code class="docutils literal notranslate"><span class="pre">CGModel.create_system()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.move_COM_to_box_center"><code class="docutils literal notranslate"><span class="pre">CGModel.move_COM_to_box_center()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.save_state"><code class="docutils literal notranslate"><span class="pre">CGModel.save_state()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.save_system"><code class="docutils literal notranslate"><span class="pre">CGModel.save_system()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.set_rigid_bodies"><code class="docutils literal notranslate"><span class="pre">CGModel.set_rigid_bodies()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_all_elec"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_all_elec()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_all_vdwl"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_all_vdwl()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_angles"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_dna_angles()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_base_pairs"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_dna_base_pairs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_bonds"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_dna_bonds()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_cross_stackings"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_dna_cross_stackings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_dihedrals"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_dna_dihedrals()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_stackings"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_dna_stackings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_native_pairs"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_native_pairs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_angles"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_protein_angles()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_bonds"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_protein_bonds()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_dihedrals"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.add_protein_dihedrals()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.append_mol"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.append_mol()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.legacy_add_dna_cross_stackings"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.legacy_add_dna_cross_stackings()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.parse_all_exclusions"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.parse_all_exclusions()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.parse_dna_exclusions"><code class="docutils literal notranslate"><span class="pre">SMOG3SPN2Model.parse_dna_exclusions()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-OpenSMOG3SPN2.forcefields.parsers.smog_parser">OpenSMOG3SPN2.forcefields.parsers</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser"><code class="docutils literal notranslate"><span class="pre">SMOGParser</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.from_atomistic_pdb"><code class="docutils literal notranslate"><span class="pre">SMOGParser.from_atomistic_pdb()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.parse_exclusions"><code class="docutils literal notranslate"><span class="pre">SMOGParser.parse_exclusions()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.parse_mol"><code class="docutils literal notranslate"><span class="pre">SMOGParser.parse_mol()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.aa_to_cg"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser.aa_to_cg()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.build_x3dna_template"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser.build_x3dna_template()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.change_sequence"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser.change_sequence()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.from_atomistic_pdb"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser.from_atomistic_pdb()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser.get_sequence()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence_list"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser.get_sequence_list()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence_list_from_cg_atoms"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser.get_sequence_list_from_cg_atoms()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.parse_mol"><code class="docutils literal notranslate"><span class="pre">DNA3SPN2Parser.parse_mol()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-OpenSMOG3SPN2.utils.helper_functions">OpenSMOG3SPN2.utils.helper_functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.atomistic_pdb_to_ca_pdb"><code class="docutils literal notranslate"><span class="pre">atomistic_pdb_to_ca_pdb()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.atomistic_pdb_to_nucleotide_pdb"><code class="docutils literal notranslate"><span class="pre">atomistic_pdb_to_nucleotide_pdb()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.build_straight_CA_chain"><code class="docutils literal notranslate"><span class="pre">build_straight_CA_chain()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.compute_rg"><code class="docutils literal notranslate"><span class="pre">compute_rg()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.fix_pdb"><code class="docutils literal notranslate"><span class="pre">fix_pdb()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.get_WC_paired_seq"><code class="docutils literal notranslate"><span class="pre">get_WC_paired_seq()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.make_mol_whole"><code class="docutils literal notranslate"><span class="pre">make_mol_whole()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.move_atoms_to_closest_pbc_image"><code class="docutils literal notranslate"><span class="pre">move_atoms_to_closest_pbc_image()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.parse_pdb"><code class="docutils literal notranslate"><span class="pre">parse_pdb()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.helper_functions.write_pdb"><code class="docutils literal notranslate"><span class="pre">write_pdb()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-OpenSMOG3SPN2.utils.chromatin_helper_functions">OpenSMOG3SPN2.utils.chromatin_helper_functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.chromatin_helper_functions.get_chromatin_rigid_bodies"><code class="docutils literal notranslate"><span class="pre">get_chromatin_rigid_bodies()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.chromatin_helper_functions.remove_histone_tail_dihedrals"><code class="docutils literal notranslate"><span class="pre">remove_histone_tail_dihedrals()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.chromatin_helper_functions.remove_histone_tail_native_pairs_and_exclusions"><code class="docutils literal notranslate"><span class="pre">remove_histone_tail_native_pairs_and_exclusions()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-OpenSMOG3SPN2.utils.shadow_map">OpenSMOG3SPN2.utils.shadow_map</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.shadow_map.find_ca_pairs_from_atomistic_pdb"><code class="docutils literal notranslate"><span class="pre">find_ca_pairs_from_atomistic_pdb()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.shadow_map.find_res_pairs_from_atomistic_pdb"><code class="docutils literal notranslate"><span class="pre">find_res_pairs_from_atomistic_pdb()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.shadow_map.get_bonded_neighbor_dict"><code class="docutils literal notranslate"><span class="pre">get_bonded_neighbor_dict()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.shadow_map.get_neighbor_pairs_and_distances"><code class="docutils literal notranslate"><span class="pre">get_neighbor_pairs_and_distances()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.shadow_map.light_is_blocked"><code class="docutils literal notranslate"><span class="pre">light_is_blocked()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.shadow_map.load_ca_pairs_from_gmx_top"><code class="docutils literal notranslate"><span class="pre">load_ca_pairs_from_gmx_top()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-OpenSMOG3SPN2.utils.insert">OpenSMOG3SPN2.utils.insert</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#OpenSMOG3SPN2.utils.insert.insert_molecules"><code class="docutils literal notranslate"><span class="pre">insert_molecules()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">OpenSMOG3SPN2</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Manual</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/manual.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="manual">
<h1>Manual<a class="headerlink" href="#manual" title="Permalink to this heading"></a></h1>
<section id="module-OpenSMOG3SPN2.forcefields.cg_model">
<span id="opensmog3spn2-forcefields"></span><h2>OpenSMOG3SPN2.forcefields<a class="headerlink" href="#module-OpenSMOG3SPN2.forcefields.cg_model" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">CGModel</span></span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel" title="Permalink to this definition"></a></dt>
<dd><p>The general class with general methods that can be inherited by any other CG model classes.</p>
<p>Each child class of CGModel has to set self.bonded_attr_names, which is used to append molecules and set rigid bodies.</p>
<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.add_plumed">
<span class="sig-name descname"><span class="pre">add_plumed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">plumed_script_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.add_plumed" title="Permalink to this definition"></a></dt>
<dd><p>Add OpenMM plumed.</p>
<section id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>plumed_script_path<span class="classifier">str</span></dt><dd><p>Input plumed script path.</p>
</dd>
</dl>
</section>
<section id="reference">
<h3>Reference<a class="headerlink" href="#reference" title="Permalink to this heading"></a></h3>
<blockquote>
<div><p>PLUMED website: <a class="reference external" href="https://www.plumed.org/">https://www.plumed.org/</a></p>
</div></blockquote>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.add_reporters">
<span class="sig-name descname"><span class="pre">add_reporters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">report_interval</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dcd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'output.dcd'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_dcd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.add_reporters" title="Permalink to this definition"></a></dt>
<dd><p>Add reporters for OpenMM simulation.</p>
<section id="id1">
<h3>Parameters<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>report_interval<span class="classifier">int</span></dt><dd><p>Report dcd and report state interval.</p>
</dd>
<dt>output_dcd<span class="classifier">str</span></dt><dd><p>Output dcd file path.</p>
</dd>
<dt>report_dcd<span class="classifier">bool</span></dt><dd><p>Whether to output dcd file.</p>
</dd>
<dt>report_state<span class="classifier">bool</span></dt><dd><p>Whether to output simulation state.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.append_mol">
<span class="sig-name descname"><span class="pre">append_mol</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_mol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.append_mol" title="Permalink to this definition"></a></dt>
<dd><p>The method can append new molecules by concatenating atoms and bonded interaction information saved in dataframes.</p>
<section id="id2">
<h3>Parameters<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>new_mol<span class="classifier">a consistent parser object or CG model object</span></dt><dd><p>The object of a new molecule including atom and bonded interaction information.</p>
</dd>
<dt>verbose<span class="classifier">bool</span></dt><dd><p>Whether to report the appended attributes.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.atoms_to_pdb">
<span class="sig-name descname"><span class="pre">atoms_to_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cg_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reset_serial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.atoms_to_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Save atoms to pdb as topology will be read from the pdb file.
PDB file is required for OpenMM simulation.</p>
<section id="id3">
<h3>Parameters<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>cg_pdb<span class="classifier">str</span></dt><dd><p>Output path for CG PDB file.</p>
</dd>
<dt>reset_serial<span class="classifier">bool</span></dt><dd><p>Reset serial to 1, 2, …</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.create_system">
<span class="sig-name descname"><span class="pre">create_system</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">top</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_pbc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_a</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_b</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_c</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_cmmotion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.create_system" title="Permalink to this definition"></a></dt>
<dd><p>Create OpenMM system for simulation. 
Need to further add forces to this OpenMM system.</p>
<section id="id4">
<h3>Parameters<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>top<span class="classifier">OpenMM Topology</span></dt><dd><p>The OpenMM topology.</p>
</dd>
<dt>use_pbc<span class="classifier">bool</span></dt><dd><p>Whether to use periodic boundary condition (PBC).</p>
</dd>
<dt>box_a<span class="classifier">float or int</span></dt><dd><p>Box length along x-axis.</p>
</dd>
<dt>box_b<span class="classifier">float or int</span></dt><dd><p>Box length along y-axis.</p>
</dd>
<dt>box_c<span class="classifier">float or int</span></dt><dd><p>Box length along z-axis.</p>
</dd>
<dt>remove_cmmotion<span class="classifier">bool</span></dt><dd><p>Whether to add CMMotionRemover to remove center of mass motions.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.move_COM_to_box_center">
<span class="sig-name descname"><span class="pre">move_COM_to_box_center</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.move_COM_to_box_center" title="Permalink to this definition"></a></dt>
<dd><p>Move center of mass (COM) to simulation box center.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.save_state">
<span class="sig-name descname"><span class="pre">save_state</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">state_xml</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'state.xml'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.save_state" title="Permalink to this definition"></a></dt>
<dd><p>Save simulation state to readable xml format.</p>
<section id="id5">
<h3>Parameters<a class="headerlink" href="#id5" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>state_xml<span class="classifier">str</span></dt><dd><p>Output state xml file path.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.save_system">
<span class="sig-name descname"><span class="pre">save_system</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">system_xml</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'system.xml'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.save_system" title="Permalink to this definition"></a></dt>
<dd><p>Save system to readable xml format.</p>
<section id="id6">
<h3>Parameters<a class="headerlink" href="#id6" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>system_xml<span class="classifier">str</span></dt><dd><p>Output system xml file path.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.cg_model.CGModel.set_rigid_bodies">
<span class="sig-name descname"><span class="pre">set_rigid_bodies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rigid_coord</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rigid_bodies</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_unchanged</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel.set_rigid_bodies" title="Permalink to this definition"></a></dt>
<dd><p>Set rigid bodies and remove bonded interactions within the same rigid body.</p>
<p>Directly run <cite>createRigidBodies</cite> if the user does not want to change any bonded interactions or exclusions.</p>
<section id="id7">
<h3>Parameters<a class="headerlink" href="#id7" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>rigid_coord<span class="classifier">Quantity, shape = (n_atoms, 3)</span></dt><dd><p>Atom coordinates to build rigid body restrictions.</p>
</dd>
<dt>rigid_bodies<span class="classifier">list </span></dt><dd><p>A list includes many sublists. Atoms in each sublist are recognized as one rigid body.</p>
</dd>
<dt>keep_unchanged<span class="classifier">list</span></dt><dd><p>A list including attribute names that user intends to keep unchanged.</p>
</dd>
</dl>
</section>
</dd></dl>

</dd></dl>

<span class="target" id="module-OpenSMOG3SPN2.forcefields.smog_3spn2_model"></span><dl class="py class">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SMOG3SPN2Model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dna_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'B_curved'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">OpenCLPatch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_parse_config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#OpenSMOG3SPN2.forcefields.cg_model.CGModel" title="OpenSMOG3SPN2.forcefields.cg_model.CGModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">CGModel</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Mixin3SPN2ConfigParser</span></code></p>
<p>The class for SMOG+3SPN2 model. 
To ensure this model works properly, please ensure two neighboring ssDNA chains do not share same chainID.</p>
<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_all_elec">
<span class="sig-name descname"><span class="pre">add_all_elec</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">salt_conc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">Quantity(value=150,</span> <span class="pre">unit=millimolar)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temperature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">Quantity(value=300,</span> <span class="pre">unit=kelvin)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elec_DD_charge_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_DD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">Quantity(value=5,</span> <span class="pre">unit=nanometer)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_PP_PD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">Quantity(value=3.141504539,</span> <span class="pre">unit=nanometer)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dielectric_PP_PD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">78</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_all_elec" title="Permalink to this definition"></a></dt>
<dd><p>Add all the electrostatic interactions.</p>
<section id="id8">
<h3>Parameters<a class="headerlink" href="#id8" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>salt_conc<span class="classifier">Quantity</span></dt><dd><p>Monovalent salt concentration.</p>
</dd>
<dt>temperature<span class="classifier">Quantity</span></dt><dd><p>Temperature.</p>
</dd>
<dt>elec_DD_charge_scale<span class="classifier">float or int</span></dt><dd><p>Charge scale factor for DNA-DNA electrostatic interactions.</p>
</dd>
<dt>cutoff_DD<span class="classifier">Quantity</span></dt><dd><p>DNA-DNA electrostatic interaction cutoff.</p>
</dd>
<dt>cutoff_PP_PD<span class="classifier">Quantity</span></dt><dd><p>Protein-protein and protein-DNA electrostatic interaction cutoff.</p>
</dd>
<dt>dielectric_PP_PD<span class="classifier">float or int</span></dt><dd><p>Protein-protein and protein-DNA electrostatic interaction dielectric.</p>
</dd>
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_all_vdwl">
<span class="sig-name descname"><span class="pre">add_all_vdwl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">param_PP_MJ_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/home/gridsan/sliu/Projects/smog-3spn2-openmm/OpenSMOG3SPN2/forcefields/parameters/pp_MJ.csv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff_PD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">Quantity(value=1.425,</span> <span class="pre">unit=nanometer)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">11</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_all_vdwl" title="Permalink to this definition"></a></dt>
<dd><p>Add all the nonbonded Van der Waals interactions.</p>
<p>CG atom type 0-19 for amino acids. CG atom type 20-25 for DNA atoms.</p>
<section id="id9">
<h3>Parameters<a class="headerlink" href="#id9" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>param_PP_MJ<span class="classifier">str</span></dt><dd><p>Protein-protein MJ potential parameter file path.</p>
</dd>
<dt>cutoff_PD<span class="classifier">Quantity</span></dt><dd><p>Protein-DNA Van der Waals interaction cutoff distance.</p>
</dd>
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_angles">
<span class="sig-name descname"><span class="pre">add_dna_angles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_angles" title="Permalink to this definition"></a></dt>
<dd><p>Add DNA angles.</p>
<section id="id10">
<h3>Parameters<a class="headerlink" href="#id10" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_base_pairs">
<span class="sig-name descname"><span class="pre">add_dna_base_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_base_pairs" title="Permalink to this definition"></a></dt>
<dd><p>Add DNA base pair potentials.</p>
<p>Please ensure two neighboring chains in self.atoms do not share the same chainID, so that different chains can be distinguished properly.</p>
<section id="id11">
<h3>Parameters<a class="headerlink" href="#id11" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>cutoff<span class="classifier">float or int</span></dt><dd><p>Cutoff distance.</p>
</dd>
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_bonds">
<span class="sig-name descname"><span class="pre">add_dna_bonds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_bonds" title="Permalink to this definition"></a></dt>
<dd><p>Add DNA bonds.</p>
<section id="id12">
<h3>Parameters<a class="headerlink" href="#id12" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_cross_stackings">
<span class="sig-name descname"><span class="pre">add_dna_cross_stackings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_cross_stackings" title="Permalink to this definition"></a></dt>
<dd><p>New method for adding DNA cross stacking potentials. This method should be faster.</p>
<p>The previous method is slow because each (B1, S1, B2) atom group acts as both donor and acceptor. 
We accelerate by using (B1, S1, B2) atom group as either donor or acceptor. 
The group is a donor if resid(B1) + 1 == resid(B2), and an acceptor if resid(B1) - 1 == resid(B2).</p>
<p>Please ensure two neighboring chains in self.atoms do not share the same chainID, so that different chains can be distinguished properly.</p>
<p>The method should be efficient if OpenCLPatch is True.</p>
<section id="id13">
<h3>Parameters<a class="headerlink" href="#id13" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>cutoff<span class="classifier">float or int</span></dt><dd><p>Cutoff distance.</p>
</dd>
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_dihedrals">
<span class="sig-name descname"><span class="pre">add_dna_dihedrals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_dihedrals" title="Permalink to this definition"></a></dt>
<dd><p>Add DNA dihedrals.</p>
<section id="id14">
<h3>Parameters<a class="headerlink" href="#id14" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_stackings">
<span class="sig-name descname"><span class="pre">add_dna_stackings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_dna_stackings" title="Permalink to this definition"></a></dt>
<dd><p>Add DNA stackings.</p>
<section id="id15">
<h3>Parameters<a class="headerlink" href="#id15" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_native_pairs">
<span class="sig-name descname"><span class="pre">add_native_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_native_pairs" title="Permalink to this definition"></a></dt>
<dd><p>Add native pairs.</p>
<section id="id16">
<h3>Parameters<a class="headerlink" href="#id16" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_angles">
<span class="sig-name descname"><span class="pre">add_protein_angles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_angles" title="Permalink to this definition"></a></dt>
<dd><p>Add protein angles.</p>
<section id="id17">
<h3>Parameters<a class="headerlink" href="#id17" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_bonds">
<span class="sig-name descname"><span class="pre">add_protein_bonds</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_bonds" title="Permalink to this definition"></a></dt>
<dd><p>Add protein bonds.</p>
<section id="id18">
<h3>Parameters<a class="headerlink" href="#id18" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_dihedrals">
<span class="sig-name descname"><span class="pre">add_protein_dihedrals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.add_protein_dihedrals" title="Permalink to this definition"></a></dt>
<dd><p>Add protein dihedrals.</p>
<section id="id19">
<h3>Parameters<a class="headerlink" href="#id19" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.append_mol">
<span class="sig-name descname"><span class="pre">append_mol</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_mol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.append_mol" title="Permalink to this definition"></a></dt>
<dd><p>The method can append new molecules by concatenating atoms and bonded interaction information saved in dataframes. 
Please ensure two neighboring chains do not share chainID.</p>
<section id="id20">
<h3>Parameters<a class="headerlink" href="#id20" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>new_mol<span class="classifier">a consistent parser object or CG model object</span></dt><dd><p>The object of a new molecule including atom and bonded interaction information.</p>
</dd>
<dt>verbose<span class="classifier">bool</span></dt><dd><p>Whether to report the appended attributes.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.legacy_add_dna_cross_stackings">
<span class="sig-name descname"><span class="pre">legacy_add_dna_cross_stackings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.legacy_add_dna_cross_stackings" title="Permalink to this definition"></a></dt>
<dd><p>Add DNA cross stacking potentials.</p>
<p>This is the old method, which is slower, because each (B1, S1, B2) atom group acts as both donors and acceptors. 
This method is still correct, so it can be used for verifications.</p>
<p>Please ensure two neighboring chains in self.atoms do not share the same chainID, so that different chains can be distinguished properly.</p>
<p>The method should be efficient if OpenCLPatch is True.</p>
<section id="id21">
<h3>Parameters<a class="headerlink" href="#id21" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>cutoff<span class="classifier">float or int</span></dt><dd><p>Cutoff distance.</p>
</dd>
<dt>force_group<span class="classifier">int</span></dt><dd><p>Force group.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.parse_all_exclusions">
<span class="sig-name descname"><span class="pre">parse_all_exclusions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.parse_all_exclusions" title="Permalink to this definition"></a></dt>
<dd><p>Parse all the exclusions (including protein exclusions and DNA exclusions). 
Run this command before adding nonbonded interactions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.parse_dna_exclusions">
<span class="sig-name descname"><span class="pre">parse_dna_exclusions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.smog_3spn2_model.SMOG3SPN2Model.parse_dna_exclusions" title="Permalink to this definition"></a></dt>
<dd><p>Parse DNA nonbonded interaction exclusions. 
Note we parse exclusions here instead of in parser as 3SPN2 involves exclusions between atoms from different chains. 
The code should be efficient if there are many DNA chains in the system.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-OpenSMOG3SPN2.forcefields.parsers.smog_parser">
<span id="opensmog3spn2-forcefields-parsers"></span><h2>OpenSMOG3SPN2.forcefields.parsers<a class="headerlink" href="#module-OpenSMOG3SPN2.forcefields.parsers.smog_parser" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SMOGParser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atomistic_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_parse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser" title="Permalink to this definition"></a></dt>
<dd><p>SMOG protein parser.</p>
<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.from_atomistic_pdb">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_atomistic_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atomistic_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_TER</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_parse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.from_atomistic_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Initialize an HPS model protein from atomistic pdb.</p>
<section id="id22">
<h3>Parameters<a class="headerlink" href="#id22" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>atomistic_pdb<span class="classifier">str</span></dt><dd><p>Path for the input atomistic pdb file.</p>
</dd>
<dt>ca_pdb<span class="classifier">str</span></dt><dd><p>Path for the output CA pdb file.</p>
</dd>
<dt>write_TER<span class="classifier">bool</span></dt><dd><p>Whether to write TER between two chains.</p>
</dd>
<dt>default_parse<span class="classifier">bool</span></dt><dd><p>Whether to parse with default settings.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.parse_exclusions">
<span class="sig-name descname"><span class="pre">parse_exclusions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exclude12</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude13</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude14</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_native_pairs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.parse_exclusions" title="Permalink to this definition"></a></dt>
<dd><p>Parse nonbonded exclusions based on bonds, angles, dihedrals, and native pairs.</p>
<section id="id23">
<h3>Parameters<a class="headerlink" href="#id23" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>exclude12<span class="classifier">bool</span></dt><dd><p>Whether to exclude nonbonded interactions between 1-2 atom pairs.</p>
</dd>
<dt>exclude13<span class="classifier">bool</span></dt><dd><p>Whether to exclude nonbonded interactions between 1-3 atom pairs.</p>
</dd>
<dt>exclude14<span class="classifier">bool</span></dt><dd><p>Whether to exclude nonbonded interactions between 1-4 atom pairs.</p>
</dd>
<dt>exclude_native_pairs<span class="classifier">bool</span></dt><dd><p>Whether to exclude nonbonded interactions between native pairs.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.parse_mol">
<span class="sig-name descname"><span class="pre">parse_mol</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">get_native_pairs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bonded_radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_pbc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude12</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude13</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude14</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_native_pairs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mass_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'ALA':</span> <span class="pre">71.0788,</span> <span class="pre">'ARG':</span> <span class="pre">156.1875,</span> <span class="pre">'ASN':</span> <span class="pre">114.1038,</span> <span class="pre">'ASP':</span> <span class="pre">115.0886,</span> <span class="pre">'CYS':</span> <span class="pre">103.1388,</span> <span class="pre">'GLN':</span> <span class="pre">128.1307,</span> <span class="pre">'GLU':</span> <span class="pre">129.1155,</span> <span class="pre">'GLY':</span> <span class="pre">57.0519,</span> <span class="pre">'HIS':</span> <span class="pre">137.1411,</span> <span class="pre">'ILE':</span> <span class="pre">113.1594,</span> <span class="pre">'LEU':</span> <span class="pre">113.1594,</span> <span class="pre">'LYS':</span> <span class="pre">128.1741,</span> <span class="pre">'MET':</span> <span class="pre">131.1926,</span> <span class="pre">'PHE':</span> <span class="pre">147.1766,</span> <span class="pre">'PRO':</span> <span class="pre">97.1167,</span> <span class="pre">'SER':</span> <span class="pre">87.0782,</span> <span class="pre">'THR':</span> <span class="pre">101.1051,</span> <span class="pre">'TRP':</span> <span class="pre">186.2132,</span> <span class="pre">'TYR':</span> <span class="pre">163.176,</span> <span class="pre">'VAL':</span> <span class="pre">99.1326}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">charge_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'ALA':</span> <span class="pre">0.0,</span> <span class="pre">'ARG':</span> <span class="pre">1.0,</span> <span class="pre">'ASN':</span> <span class="pre">0.0,</span> <span class="pre">'ASP':</span> <span class="pre">-1.0,</span> <span class="pre">'CYS':</span> <span class="pre">0.0,</span> <span class="pre">'GLN':</span> <span class="pre">0.0,</span> <span class="pre">'GLU':</span> <span class="pre">-1.0,</span> <span class="pre">'GLY':</span> <span class="pre">0.0,</span> <span class="pre">'HIS':</span> <span class="pre">0.0,</span> <span class="pre">'ILE':</span> <span class="pre">0.0,</span> <span class="pre">'LEU':</span> <span class="pre">0.0,</span> <span class="pre">'LYS':</span> <span class="pre">1.0,</span> <span class="pre">'MET':</span> <span class="pre">0.0,</span> <span class="pre">'PHE':</span> <span class="pre">0.0,</span> <span class="pre">'PRO':</span> <span class="pre">0.0,</span> <span class="pre">'SER':</span> <span class="pre">0.0,</span> <span class="pre">'THR':</span> <span class="pre">0.0,</span> <span class="pre">'TRP':</span> <span class="pre">0.0,</span> <span class="pre">'TYR':</span> <span class="pre">0.0,</span> <span class="pre">'VAL':</span> <span class="pre">0.0}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bonded_energy_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.smog_parser.SMOGParser.parse_mol" title="Permalink to this definition"></a></dt>
<dd><p>Parse molecule.</p>
<section id="id24">
<h3>Parameters<a class="headerlink" href="#id24" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>get_native_pairs<span class="classifier">bool</span></dt><dd><p>Whether to get native pairs from atomistic pdb with shadow algorithm.</p>
</dd>
<dt>frame<span class="classifier">int</span></dt><dd><p>Frame index to compute native configuration parameters for both CG and atomistic models.</p>
</dd>
<dt>radius<span class="classifier">float or int</span></dt><dd><p>Shadow algorithm radius.</p>
</dd>
<dt>bonded_radius<span class="classifier">float or int</span></dt><dd><p>Shadow algorithm bonded radius.</p>
</dd>
<dt>cutoff<span class="classifier">float or int</span></dt><dd><p>Shadow algorithm cutoff.</p>
</dd>
</dl>
<p>box : None or np.ndarray
Specify box shape. 
Note <cite>use_pbc</cite> = False is only compatible with orthogonal box. 
If <cite>use_pbc</cite> = False, then set <cite>box</cite> as None. 
If <cite>use_pbc</cite> = True, then <cite>box</cite> = np.array([lx, ly, lz, alpha1, alpha2, alpha3]).</p>
<dl class="simple">
<dt>use_pbc<span class="classifier">bool</span></dt><dd><p>Whether to use periodic boundary condition (PBC) for searching native pairs.</p>
</dd>
<dt>exclude12<span class="classifier">bool</span></dt><dd><p>Whether to exclude nonbonded interactions between 1-2 atom pairs.</p>
</dd>
<dt>exclude13<span class="classifier">bool</span></dt><dd><p>Whether to exclude nonbonded interactions between 1-3 atom pairs.</p>
</dd>
<dt>exclude14<span class="classifier">bool</span></dt><dd><p>Whether to exclude nonbonded interactions between 1-4 atom pairs.</p>
</dd>
<dt>exclude_native_pairs<span class="classifier">bool</span></dt><dd><p>Whether to exclude nonbonded interactions between native pairs.</p>
</dd>
<dt>mass_dict<span class="classifier">dict</span></dt><dd><p>Mass dictionary.</p>
</dd>
<dt>charge_dict<span class="classifier">dict</span></dt><dd><p>Charge dictionary.</p>
</dd>
<dt>bonded_energy_scale<span class="classifier">float or int</span></dt><dd><p>Bonded energy additional scale factor.</p>
</dd>
</dl>
</section>
</dd></dl>

</dd></dl>

<span class="target" id="module-OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser"></span><dl class="py class">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DNA3SPN2Parser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cg_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dna_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'B_curved'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser" title="Permalink to this definition"></a></dt>
<dd><p>DNA 3SPN2 parser. 
For B-curved DNA, the parser works best for a single strand ssDNA or WC-paired dsDNA. 
Please ensure the parsed DNA has unique chainID for each chain.</p>
<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.aa_to_cg">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">aa_to_cg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_atoms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PSB_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.aa_to_cg" title="Permalink to this definition"></a></dt>
<dd><p>Convert DNA all-atom structure to CG structure. 
Both input and output structures are saved as pandas dataframe. 
Notably, this function gives each residue in the output CG model a unique resSeq (index starts from 0). 
Using unique resSeq can be helpful when comparing with x3dna built template as chainID may be different. 
This code is long and specific to 3SPN2 model, so we put it here instead of in helper_functions.py.</p>
<section id="id25">
<h3>Parameters<a class="headerlink" href="#id25" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>aa_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>Input all-atom structure.</p>
</dd>
<dt>PSB_order<span class="classifier">bool</span></dt><dd><p>Whether to ensure CG atom order in each nucleotide is P-S-B.</p>
</dd>
</dl>
</section>
<section id="returns">
<h3>Returns<a class="headerlink" href="#returns" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>cg_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>Output CG structure.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.build_x3dna_template">
<span class="sig-name descname"><span class="pre">build_x3dna_template</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">temp_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dna'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.build_x3dna_template" title="Permalink to this definition"></a></dt>
<dd><p>Build template DNA structure with x3dna. 
We do not need to specify whether to use PSB order, since template atom order is aligned with self.atoms.
If self.atoms is one dsDNA molecule and the target sequence is W-C paired, x3dna input sequence is the sequence of the first ssDNA. Else, use the full sequence of the DNA as x3dna input sequence.</p>
<section id="id26">
<h3>Parameters<a class="headerlink" href="#id26" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>temp_name<span class="classifier">str</span></dt><dd><p>Name for built template files.</p>
</dd>
</dl>
</section>
<section id="id27">
<h3>Returns<a class="headerlink" href="#id27" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>temp_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>X3DNA built CG DNA template structure.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.change_sequence">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">change_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cg_atoms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.change_sequence" title="Permalink to this definition"></a></dt>
<dd><p>Change DNA sequence. 
Note sequence includes the sequence of all the ssDNA chains.</p>
<section id="id28">
<h3>Parameters<a class="headerlink" href="#id28" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>cg_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>DNA CG structure.</p>
</dd>
<dt>sequence<span class="classifier">str</span></dt><dd><p>New DNA sequence.</p>
</dd>
</dl>
</section>
<section id="id29">
<h3>Returns<a class="headerlink" href="#id29" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>new_cg_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>DNA CG structure with the new sequence.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.from_atomistic_pdb">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_atomistic_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atomistic_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cg_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PSB_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dna_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'B_curved'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_parse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_from_x3dna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dna'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.from_atomistic_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Create object from atomistic pdb file. 
Ensure each chain in input pdb_file has unique chainID. 
If a new sequence is provided, then the CG DNA topology is from input pdb file, but using the new sequence.</p>
<section id="id30">
<h3>Parameters<a class="headerlink" href="#id30" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>atomistic_pdb<span class="classifier">str</span></dt><dd><p>Path for the input atomistic pdb file.</p>
</dd>
<dt>cg_pdb<span class="classifier">str</span></dt><dd><p>Path for the output CG pdb file.</p>
</dd>
<dt>PSB_order<span class="classifier">bool</span></dt><dd><p>Whether to ensure CG atom order in each nucleotide is P-S-B.</p>
</dd>
<dt>new_sequence<span class="classifier">None or str</span></dt><dd><p>The new full DNA sequence. 
If None, keep the original sequence.</p>
</dd>
<dt>dna_type<span class="classifier">str</span></dt><dd><p>DNA type.</p>
</dd>
<dt>default_parse<span class="classifier">bool</span></dt><dd><p>Whether to parse molecule with default settings.</p>
</dd>
<dt>temp_from_x3dna<span class="classifier">bool</span></dt><dd><p>Whether to use template built by x3dna. 
If False, use the input pdb file as the template.</p>
</dd>
<dt>temp_name<span class="classifier">str</span></dt><dd><p>Template file names.</p>
</dd>
<dt>input_temp<span class="classifier">None or pd.DataFrame</span></dt><dd><p>Additional input template.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence">
<span class="sig-name descname"><span class="pre">get_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence" title="Permalink to this definition"></a></dt>
<dd><p>Get all ssDNA sequence from CG atoms as a string. 
Note each chain has to possess a unique chainID.</p>
<section id="id31">
<h3>Returns<a class="headerlink" href="#id31" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sequence<span class="classifier">str</span></dt><dd><p>DNA sequence.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence_list">
<span class="sig-name descname"><span class="pre">get_sequence_list</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence_list" title="Permalink to this definition"></a></dt>
<dd><p>Get all ssDNA sequence from self.atoms as a list. 
Note each chain has to possess a unique chainID.</p>
<section id="id32">
<h3>Returns<a class="headerlink" href="#id32" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sequence_list<span class="classifier">list</span></dt><dd><p>A list including multiple strings. Each string is the sequence of one chain.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence_list_from_cg_atoms">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_sequence_list_from_cg_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cg_atoms</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.get_sequence_list_from_cg_atoms" title="Permalink to this definition"></a></dt>
<dd><p>Get all ssDNA sequence from CG atoms as a list. 
Since we use <cite>groupby</cite> to group by chainID, each chain has to possess a unique chainID.</p>
<section id="id33">
<h3>Parameters<a class="headerlink" href="#id33" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>cg_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>CG DNA atoms.</p>
</dd>
</dl>
</section>
<section id="id34">
<h3>Returns<a class="headerlink" href="#id34" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sequence_list<span class="classifier">list</span></dt><dd><p>A list including multiple strings. Each string is the sequence of one chain.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.parse_mol">
<span class="sig-name descname"><span class="pre">parse_mol</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">temp_from_x3dna</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dna'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_temp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.forcefields.parsers.dna_3spn2_parser.DNA3SPN2Parser.parse_mol" title="Permalink to this definition"></a></dt>
<dd><p>Parse molecule with given template.</p>
<section id="id35">
<h3>Parameters<a class="headerlink" href="#id35" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>temp_from_x3dna<span class="classifier">bool</span></dt><dd><p>Whether to use x3dna template.</p>
</dd>
<dt>temp_name<span class="classifier">str</span></dt><dd><p>Name for x3dna built template files.</p>
</dd>
<dt>input_temp<span class="classifier">None or pd.DataFrame</span></dt><dd><p>Additional input template. Use input_temp only <cite>if (not ((self.dna_type == ‘B_curved’) and temp_from_x3dna)) and (input_temp is not None)</cite>.</p>
</dd>
</dl>
</section>
</dd></dl>

</dd></dl>

</section>
<section id="module-OpenSMOG3SPN2.utils.helper_functions">
<span id="opensmog3spn2-utils-helper-functions"></span><h2>OpenSMOG3SPN2.utils.helper_functions<a class="headerlink" href="#module-OpenSMOG3SPN2.utils.helper_functions" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.atomistic_pdb_to_ca_pdb">
<span class="sig-name descname"><span class="pre">atomistic_pdb_to_ca_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atomistic_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_TER</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.atomistic_pdb_to_ca_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Convert atomistic pdb to protein CA pdb.</p>
<section id="id36">
<h3>Parameters<a class="headerlink" href="#id36" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>atomistic_pdb<span class="classifier">str</span></dt><dd><p>Path for the atomistic pdb file.</p>
</dd>
<dt>ca_pdb<span class="classifier">str</span></dt><dd><p>Output path for the CA pdb file.</p>
</dd>
<dt>write_TER<span class="classifier">bool</span></dt><dd><p>Whether to write TER between two chains.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.atomistic_pdb_to_nucleotide_pdb">
<span class="sig-name descname"><span class="pre">atomistic_pdb_to_nucleotide_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atomistic_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cg_nucleotide_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_TER</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.atomistic_pdb_to_nucleotide_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Convert atomistic pdb to DNA nucleotide pdb (i.e. one CG bead per nucleotide). 
The position of each CG nucleotide bead is the geometric center of all the nucleotide atoms in the pdb.</p>
<section id="id37">
<h3>Parameters<a class="headerlink" href="#id37" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>atomistic_pdb<span class="classifier">str</span></dt><dd><p>Path for the atomistic pdb file.</p>
</dd>
<dt>cg_nucleotide_pdb<span class="classifier">str</span></dt><dd><p>Output path for the CG pdb file.</p>
</dd>
<dt>write_TER<span class="classifier">bool</span></dt><dd><p>Whether to write TER between two chains.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.build_straight_CA_chain">
<span class="sig-name descname"><span class="pre">build_straight_CA_chain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r0</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.38</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.build_straight_CA_chain" title="Permalink to this definition"></a></dt>
<dd><p>Build a straight portein CA atom chain with given sequence.</p>
<section id="id38">
<h3>Parameters<a class="headerlink" href="#id38" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sequence<span class="classifier">str</span></dt><dd><p>Protein chain sequence (1 letter for each amino acid).</p>
</dd>
<dt>r0: float or int</dt><dd><p>Distance in unit nm between two neighboring CA atoms.</p>
</dd>
</dl>
</section>
<section id="id39">
<h3>Returns<a class="headerlink" href="#id39" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>df_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>A pandas dataframe includes atom information.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.compute_rg">
<span class="sig-name descname"><span class="pre">compute_rg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coord</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mass</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.compute_rg" title="Permalink to this definition"></a></dt>
<dd><p>Compute radius of gyration.</p>
<section id="id40">
<h3>Parameters<a class="headerlink" href="#id40" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>coord<span class="classifier">np.ndarray, shape = (n_atoms, 3)</span></dt><dd><p>Input atom coordinates.</p>
</dd>
<dt>mass<span class="classifier">np.ndarray, shape = (n_atoms,)</span></dt><dd><p>Input atom mass.</p>
</dd>
</dl>
</section>
<section id="id41">
<h3>Returns<a class="headerlink" href="#id41" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>rg<span class="classifier">float</span></dt><dd><p>Radius of gyration.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.fix_pdb">
<span class="sig-name descname"><span class="pre">fix_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pdb_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.fix_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Fix pdb file with pdbfixer. The fixed structure is output as pandas dataframe. 
This function is adapted from open3SPN2.</p>
<section id="id42">
<h3>Parameters<a class="headerlink" href="#id42" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>pdb_file<span class="classifier">str</span></dt><dd><p>PDB file path.</p>
</dd>
</dl>
</section>
<section id="id43">
<h3>Returns<a class="headerlink" href="#id43" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>Output fixed structure.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.get_WC_paired_seq">
<span class="sig-name descname"><span class="pre">get_WC_paired_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.get_WC_paired_seq" title="Permalink to this definition"></a></dt>
<dd><p>Get WC-paired sequence.</p>
<section id="id44">
<h3>Parameters<a class="headerlink" href="#id44" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sequence<span class="classifier">str</span></dt><dd><p>Input sequence.</p>
</dd>
</dl>
</section>
<section id="id45">
<h3>Returns<a class="headerlink" href="#id45" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>paired_sequence<span class="classifier">str</span></dt><dd><p>W-C paired sequence.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.make_mol_whole">
<span class="sig-name descname"><span class="pre">make_mol_whole</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coord</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_c</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.make_mol_whole" title="Permalink to this definition"></a></dt>
<dd><p>Use this function to make the molecule whole. 
This is useful to recover the whole molecule if it is split by box boundary. 
The position of the first atom is kept. 
The other atoms are moved so that each one is in the closest periodic image relative to the previous one.</p>
<section id="id46">
<h3>Parameters<a class="headerlink" href="#id46" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>coord<span class="classifier">np.ndarray, shape = (n_atoms, 3)</span></dt><dd><p>Input atom coordinates.</p>
</dd>
<dt>box_a<span class="classifier">float or int</span></dt><dd><p>Box length along x-axis.</p>
</dd>
<dt>box_b<span class="classifier">float or int</span></dt><dd><p>Box length along y-axis.</p>
</dd>
<dt>box_c<span class="classifier">float or int</span></dt><dd><p>Box length along z-axis.</p>
</dd>
</dl>
</section>
<section id="id47">
<h3>Returns<a class="headerlink" href="#id47" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>new_coord<span class="classifier">np.ndarray, shape = (n_atoms, 3)</span></dt><dd><p>Output atom coordinates.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.move_atoms_to_closest_pbc_image">
<span class="sig-name descname"><span class="pre">move_atoms_to_closest_pbc_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coord</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_point</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_b</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box_c</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.move_atoms_to_closest_pbc_image" title="Permalink to this definition"></a></dt>
<dd><p>Move the atom coordinates to the periodic image closest to ref_point.</p>
<section id="id48">
<h3>Parameters<a class="headerlink" href="#id48" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>coord<span class="classifier">np.ndarray, shape = (n_atoms, 3)</span></dt><dd><p>Input atom coordinates.</p>
</dd>
<dt>ref_point<span class="classifier">np.ndarray, shape = (3,)</span></dt><dd><p>The reference point coordinate.</p>
</dd>
<dt>box_a<span class="classifier">float or int</span></dt><dd><p>Box length along x-axis.</p>
</dd>
<dt>box_b<span class="classifier">float or int</span></dt><dd><p>Box length along y-axis.</p>
</dd>
<dt>box_c<span class="classifier">float or int</span></dt><dd><p>Box length along z-axis.</p>
</dd>
</dl>
</section>
<section id="id49">
<h3>Returns<a class="headerlink" href="#id49" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>new_coord<span class="classifier">np.ndarray, shape = (n_atoms, 3)</span></dt><dd><p>Output atom coordinates.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.parse_pdb">
<span class="sig-name descname"><span class="pre">parse_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pdb_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.parse_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Load pdb file as pandas dataframe.</p>
<section id="id50">
<h3>Parameters<a class="headerlink" href="#id50" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>pdb_file<span class="classifier">str</span></dt><dd><p>Path for the pdb file.</p>
</dd>
</dl>
</section>
<section id="id51">
<h3>Returns<a class="headerlink" href="#id51" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>pdb_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>A pandas dataframe includes atom information.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.helper_functions.write_pdb">
<span class="sig-name descname"><span class="pre">write_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pdb_atoms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdb_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_TER</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.helper_functions.write_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Write pandas dataframe to pdb file.</p>
<section id="id52">
<h3>Parameters<a class="headerlink" href="#id52" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>pdb_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>A pandas dataframe includes atom information.</p>
</dd>
<dt>pdb_file<span class="classifier">str</span></dt><dd><p>Output path for the pdb file.</p>
</dd>
<dt>write_TER<span class="classifier">bool</span></dt><dd><p>Whether to write TER between two chains.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-OpenSMOG3SPN2.utils.chromatin_helper_functions">
<span id="opensmog3spn2-utils-chromatin-helper-functions"></span><h2>OpenSMOG3SPN2.utils.chromatin_helper_functions<a class="headerlink" href="#module-OpenSMOG3SPN2.utils.chromatin_helper_functions" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.chromatin_helper_functions.get_chromatin_rigid_bodies">
<span class="sig-name descname"><span class="pre">get_chromatin_rigid_bodies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_nucl</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrl</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_rigid_bp_per_nucl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">73</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.chromatin_helper_functions.get_chromatin_rigid_bodies" title="Permalink to this definition"></a></dt>
<dd><p>Get chromatin rigid bodies. 
The chromatin should possess uniform linker length without additional linkers on both ends.</p>
<section id="id53">
<h3>Parameters<a class="headerlink" href="#id53" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>n_nucl<span class="classifier">int</span></dt><dd><p>Nucleosome number.</p>
</dd>
<dt>nrl<span class="classifier">int</span></dt><dd><p>Nucleosome repeat length.</p>
</dd>
<dt>n_flexible_bp_per_nucl<span class="classifier">int</span></dt><dd><p>The number of flexible nucleosomal base pairs for each nucleosome.</p>
</dd>
</dl>
</section>
<section id="id54">
<h3>Returns<a class="headerlink" href="#id54" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>rigid_bodies<span class="classifier">list</span></dt><dd><p>List of rigid bodies.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.chromatin_helper_functions.remove_histone_tail_dihedrals">
<span class="sig-name descname"><span class="pre">remove_histone_tail_dihedrals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_dihedrals</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.chromatin_helper_functions.remove_histone_tail_dihedrals" title="Permalink to this definition"></a></dt>
<dd><p>Remove histone tail dihedral potentials from a single histone.</p>
<p>A dihedral potential is removed if at least one atom involved is within histone tail.</p>
<section id="id55">
<h3>Parameters<a class="headerlink" href="#id55" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>df_dihedrals<span class="classifier">pd.DataFrame</span></dt><dd><p>Dihedral potential. This should only include histones. 
Each histone includes 974 CA atoms, and there should be 974*n CG atoms in all (n is the number of histones).</p>
</dd>
</dl>
</section>
<section id="id56">
<h3>Returns<a class="headerlink" href="#id56" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>new_df_dihedrals<span class="classifier">pd.DataFrame</span></dt><dd><p>New dihedral potential.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.chromatin_helper_functions.remove_histone_tail_native_pairs_and_exclusions">
<span class="sig-name descname"><span class="pre">remove_histone_tail_native_pairs_and_exclusions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_native_pairs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_exclusions</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.chromatin_helper_functions.remove_histone_tail_native_pairs_and_exclusions" title="Permalink to this definition"></a></dt>
<dd><p>Remove histone tail native pair potentials and corresponding exclusions from a single histone. 
A native pair potential is removed if at least one atom involved is within histone tail. 
The corresponding exclusions should also be removed if the native pair is removed.
Also remove native pairs between atoms from different histones.</p>
<section id="id57">
<h3>Parameters<a class="headerlink" href="#id57" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>df_native_pairs<span class="classifier">pd.DataFrame</span></dt><dd><p>Native pair potential. This should only include histones. 
Each histone includes 974 CA atoms, and there should be 974*n CG atoms in all (n is the number of histones).</p>
</dd>
<dt>df_exclusions<span class="classifier">pd.DataFrame</span></dt><dd><p>Nonbonded exclusions.</p>
</dd>
</dl>
</section>
<section id="id58">
<h3>Returns<a class="headerlink" href="#id58" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>new_df_native_pairs<span class="classifier">pd.DataFrame</span></dt><dd><p>New native pair potential.</p>
</dd>
<dt>new_df_exclusions<span class="classifier">pd.DataFrame</span></dt><dd><p>New nonbonded exclusions.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-OpenSMOG3SPN2.utils.shadow_map">
<span id="opensmog3spn2-utils-shadow-map"></span><h2>OpenSMOG3SPN2.utils.shadow_map<a class="headerlink" href="#module-OpenSMOG3SPN2.utils.shadow_map" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.shadow_map.find_ca_pairs_from_atomistic_pdb">
<span class="sig-name descname"><span class="pre">find_ca_pairs_from_atomistic_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atomistic_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bonded_radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_pbc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.shadow_map.find_ca_pairs_from_atomistic_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Find protein CA atom pairs whose residues are in contact.</p>
<section id="id59">
<h3>Parameters<a class="headerlink" href="#id59" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>atomistic_pdb<span class="classifier">str</span></dt><dd><p>Path for the atomistic pdb file.</p>
</dd>
<dt>frame<span class="classifier">int</span></dt><dd><p>PDB frame index used for searching pairs.</p>
</dd>
<dt>radius<span class="classifier">float or int</span></dt><dd><p>Shadow radius for candidate pair atoms and blocking atoms if not bonded to any pair atom.</p>
</dd>
<dt>bonded_radius<span class="classifier">float or int</span></dt><dd><p>Shadow radius if the candidate blocking atom is directly bonded to any candidate pair atom.</p>
</dd>
<dt>cutoff: float or int</dt><dd><p>Cutoff distance for searching neighboring atom pairs.</p>
</dd>
<dt>box<span class="classifier">None or np.ndarray</span></dt><dd><p>Specify box shape. 
Note <cite>use_pbc</cite> = False is only compatible with orthogonal box. 
If <cite>use_pbc</cite> = False, then set <cite>box</cite> as None. 
If <cite>use_pbc</cite> = True, then <cite>box</cite> = np.array([lx, ly, lz, alpha1, alpha2, alpha3]).</p>
</dd>
<dt>use_pbc<span class="classifier">bool</span></dt><dd><p>Whether to use periodic boundary condition (PBC).</p>
</dd>
</dl>
</section>
<section id="id60">
<h3>Returns<a class="headerlink" href="#id60" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>df_ca_pairs<span class="classifier">pd.DataFrame</span></dt><dd><p>Pandas DataFrame including CA atom pair indices and distances.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.shadow_map.find_res_pairs_from_atomistic_pdb">
<span class="sig-name descname"><span class="pre">find_res_pairs_from_atomistic_pdb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atomistic_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bonded_radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_pbc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.shadow_map.find_res_pairs_from_atomistic_pdb" title="Permalink to this definition"></a></dt>
<dd><p>Find native pairs between residues following the shadow algorithm.
They algorithm only searches native pairs between residues that do not have 1-2, 1-3, or 1-4 interactions. 
If two heavy atoms from different residues are in contact, then the two residues are in contact.</p>
<section id="id61">
<h3>Parameters<a class="headerlink" href="#id61" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>atomistic_pdb<span class="classifier">str</span></dt><dd><p>Path for the atomistic pdb file.</p>
</dd>
<dt>frame<span class="classifier">int</span></dt><dd><p>PDB frame index used for searching pairs.</p>
</dd>
<dt>radius<span class="classifier">float or int</span></dt><dd><p>Shadow radius for candidate pair atoms and blocking atoms if not bonded to any pair atom.</p>
</dd>
<dt>bonded_radius<span class="classifier">float or int</span></dt><dd><p>Shadow radius if the candidate blocking atom is directly bonded to any candidate pair atom.</p>
</dd>
<dt>cutoff: float or int</dt><dd><p>Cutoff distance for searching neighboring atom pairs.</p>
</dd>
<dt>box<span class="classifier">None or np.ndarray</span></dt><dd><p>Specify box shape. 
Note <cite>use_pbc</cite> = False is only compatible with orthogonal box. 
If <cite>use_pbc</cite> = False, then set <cite>box</cite> as None. 
If <cite>use_pbc</cite> = True, then <cite>box</cite> = np.array([lx, ly, lz, alpha1, alpha2, alpha3]).</p>
</dd>
<dt>use_pbc<span class="classifier">bool</span></dt><dd><p>Whether to use periodic boundary condition (PBC).</p>
</dd>
</dl>
</section>
<section id="id62">
<h3>Returns<a class="headerlink" href="#id62" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>res_pairs<span class="classifier">np.ndarray, shape = (n_pairs, 2)</span></dt><dd><p>Residue pairs found by the shadow algorithm.</p>
</dd>
<dt>df_atoms<span class="classifier">pd.DataFrame</span></dt><dd><p>Pandas DataFrame including atom information from file <cite>atomistic_pdb</cite>.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.shadow_map.get_bonded_neighbor_dict">
<span class="sig-name descname"><span class="pre">get_bonded_neighbor_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atomistic_pdb</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.shadow_map.get_bonded_neighbor_dict" title="Permalink to this definition"></a></dt>
<dd><p>Find atoms that are directly connected by bonds based on MDTraj.</p>
<section id="id63">
<h3>Parameters<a class="headerlink" href="#id63" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>atomistic_pdb<span class="classifier">str</span></dt><dd><p>Path for the atomistic pdb file.</p>
</dd>
</dl>
</section>
<section id="id64">
<h3>Returns<a class="headerlink" href="#id64" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>bonded_neighbor_dict<span class="classifier">dict</span></dt><dd><p>A dictionary that the value is a list including all the atoms bonded to the atom of index key.</p>
</dd>
</dl>
</section>
<section id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this heading"></a></h3>
<p><a class="reference external" href="https://mdtraj.org/1.9.4/api/generated/mdtraj.Topology.html">https://mdtraj.org/1.9.4/api/generated/mdtraj.Topology.html</a></p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.shadow_map.get_neighbor_pairs_and_distances">
<span class="sig-name descname"><span class="pre">get_neighbor_pairs_and_distances</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coord</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_pbc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.shadow_map.get_neighbor_pairs_and_distances" title="Permalink to this definition"></a></dt>
<dd><p>Find spatially neighboring atom pairs within cutoff range based on MDAnalysis.</p>
<section id="id65">
<h3>Parameters<a class="headerlink" href="#id65" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>coord<span class="classifier">np.ndarray, shape = (n_atoms, 3)</span></dt><dd><p>Input atom coordinates.</p>
</dd>
<dt>cutoff<span class="classifier">float or int (default = 0.6)</span></dt><dd><p>Cutoff distance for searching neighboring atom pairs. 
Note <cite>coord</cite> and <cite>cutoff</cite> share the same length unit.</p>
</dd>
<dt>box<span class="classifier">None or np.ndarray</span></dt><dd><p>Specify box shape. 
Note <cite>use_pbc</cite> = False is only compatible with orthogonal box. 
If <cite>use_pbc</cite> = False, then set <cite>box</cite> as None. 
If <cite>use_pbc</cite> = True, then <cite>box</cite> = np.array([lx, ly, lz, alpha1, alpha2, alpha3]).</p>
</dd>
<dt>use_pbc<span class="classifier">bool</span></dt><dd><p>Whether to use periodic boundary condition (PBC).</p>
</dd>
</dl>
</section>
<section id="id66">
<h3>Returns<a class="headerlink" href="#id66" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>neighbor_pairs<span class="classifier">np.ndarray, shape = (n_pairs, 2)</span></dt><dd><p>Neighboring atom pairs.</p>
</dd>
<dt>neighbor_pair_distances: np.ndarray, shape = (N,)</dt><dd><p>Neighboring atom pair distances.</p>
</dd>
</dl>
</section>
<section id="id67">
<h3>References<a class="headerlink" href="#id67" title="Permalink to this heading"></a></h3>
<p><a class="reference external" href="https://docs.mdanalysis.org/1.1.1/documentation_pages/lib/nsgrid.html">https://docs.mdanalysis.org/1.1.1/documentation_pages/lib/nsgrid.html</a></p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.shadow_map.light_is_blocked">
<span class="sig-name descname"><span class="pre">light_is_blocked</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d13</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">d23</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r3</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.shadow_map.light_is_blocked" title="Permalink to this definition"></a></dt>
<dd><p>Check if the light from atom1 to atom2 is blocked by atom3.</p>
<section id="id68">
<h3>Parameters<a class="headerlink" href="#id68" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>d12: float or int, positive</dt><dd><p>The distance between atom1 and atom2.</p>
</dd>
<dt>d13: float or int, positive</dt><dd><p>The distance between atom1 and atom3.</p>
</dd>
<dt>d23: float or int, positive</dt><dd><p>The distance between atom2 and atom3.</p>
</dd>
<dt>r2: float or int, non-negative</dt><dd><p>The radius of atom2. 
Require r2 &lt;= d12.</p>
</dd>
<dt>r3: float or int, non-negative</dt><dd><p>The radius of atom3. 
Require r3 &lt;= d13.</p>
</dd>
</dl>
</section>
<section id="id69">
<h3>Returns<a class="headerlink" href="#id69" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>flag<span class="classifier">bool</span></dt><dd><p>Whether the light is blocked.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.shadow_map.load_ca_pairs_from_gmx_top">
<span class="sig-name descname"><span class="pre">load_ca_pairs_from_gmx_top</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">top_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ca_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_pbc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.shadow_map.load_ca_pairs_from_gmx_top" title="Permalink to this definition"></a></dt>
<dd><p>Load native pairs from GROMACS topology file. 
Note in GROMACS, atom indices start from 1, while in OpenMM, atom indices start from 0.</p>
<section id="id70">
<h3>Parameters<a class="headerlink" href="#id70" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>top_file<span class="classifier">str</span></dt><dd><p>Path for the GROMACS topology file.</p>
</dd>
<dt>ca_pdb<span class="classifier">str</span></dt><dd><p>Path for the CA model PDB file.</p>
</dd>
<dt>frame: int</dt><dd><p>PDB frame index used for computing pair distances.</p>
</dd>
<dt>use_pbc<span class="classifier">bool</span></dt><dd><p>Whether to use periodic boundary condition (PBC).</p>
</dd>
</dl>
</section>
<section id="id71">
<h3>Returns<a class="headerlink" href="#id71" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>df_ca_pairs<span class="classifier">pd.DataFrame</span></dt><dd><p>Pandas DataFrame including CA atom pair indices and distances.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-OpenSMOG3SPN2.utils.insert">
<span id="opensmog3spn2-utils-insert"></span><h2>OpenSMOG3SPN2.utils.insert<a class="headerlink" href="#module-OpenSMOG3SPN2.utils.insert" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="OpenSMOG3SPN2.utils.insert.insert_molecules">
<span class="sig-name descname"><span class="pre">insert_molecules</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_pdb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_mol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">radius</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">existing_pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_n_attempts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[100,</span> <span class="pre">100,</span> <span class="pre">100]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FastNS'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reset_serial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#OpenSMOG3SPN2.utils.insert.insert_molecules" title="Permalink to this definition"></a></dt>
<dd><p>Insert multiple copies of given molecule into an existing pdb or a new empty box. 
Note all the length parameter units are nm, though coordinates in pdb are in unit angstroms. 
Currently this function only supports using orthogonal box.</p>
<section id="id72">
<h3>Parameters<a class="headerlink" href="#id72" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>new_pdb<span class="classifier">str</span></dt><dd><p>Path of the new pdb file to be inserted.</p>
</dd>
<dt>output_pdb<span class="classifier">str</span></dt><dd><p>Path of the output pdb file.</p>
</dd>
<dt>n_mol<span class="classifier">int</span></dt><dd><p>Number of copies of the new molecule to be inserted.</p>
</dd>
<dt>radius<span class="classifier">float or int</span></dt><dd><p>Radius for each atom in unit nm. Insertion ensures the distance between any two atoms are larger than 2*radius under PBC condition.</p>
</dd>
<dt>existing_pdb<span class="classifier">None or str</span></dt><dd><p>Path of an existing pdb file. If None, the molecules will be inserted into a new empty box, otherwise, inserted into a copy of existing_pdb.</p>
</dd>
<dt>max_n_attempts<span class="classifier">int</span></dt><dd><p>Maximal number of attempts to insert. Ensure this value no smaller than n_mol.</p>
</dd>
<dt>box<span class="classifier">list or numpy array, shape = (3,)</span></dt><dd><p>New box size in unit nm.</p>
</dd>
<dt>method<span class="classifier">str</span></dt><dd><p>Which method to use for detecting if the inserted molecule has contact with existing atoms. 
This parameter can be ‘FastNS’ or ‘distance_array’. 
FastNS should be faster than distance_array, especially for large systems.</p>
</dd>
<dt>reset_serial<span class="classifier">bool</span></dt><dd><p>Whether to reset serial to 0, 1, …, N - 1.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to OpenSMOG3SPN2’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Zhang group.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>